From 42d7432e76c9ddb2d17ca829b930ca86095938bc Mon Sep 17 00:00:00 2001
From: Gu Chao <gux.chao@intel.com>
Date: Tue, 28 Jan 2020 14:56:43 +0800
Subject: [PATCH] DRV_DPM_SW-143: DPM fix hash with Linux kernel API

- Fix reported bug by PON traffic test
- reverts commit a8be3a41674429143a96420f1b51789aea834293.
---
 drivers/net/datapath/dpm/datapath_instance.c | 14 ++------------
 drivers/net/datapath/dpm/datapath_instance.h |  6 ++----
 drivers/net/datapath/dpm/datapath_swdev.c    | 10 ++++++----
 drivers/net/datapath/dpm/datapath_swdev.h    |  6 +-----
 4 files changed, 11 insertions(+), 25 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_instance.c b/drivers/net/datapath/dpm/datapath_instance.c
index 4a8980513515..339a1f035c2c 100644
--- a/drivers/net/datapath/dpm/datapath_instance.c
+++ b/drivers/net/datapath/dpm/datapath_instance.c
@@ -169,14 +169,7 @@ struct dp_hw_cap *match_hw_cap(struct dp_inst_info *info, u32 flag)
  */
 u32 dp_dev_hash(struct net_device *dev, char *subif_name)
 {
-	unsigned long index;
-
-	if (!dev)
-		index = (unsigned long)subif_name;
-	else
-		index = (unsigned long)dev;
-	/*Note: it is 4K alignment. Need tune later */
-	return (u32)((index >> DP_DEV_HASH_SHIFT) % DP_DEV_HASH_SIZE);
+	return hash_ptr(dev ? dev : subif_name, DP_DEV_HASH_BIT_LENGTH);
 }
 
 struct dp_dev *dp_dev_lookup(struct hlist_head *head,
@@ -379,10 +372,7 @@ int dp_inst_del_dev(struct net_device *dev, char *subif_name, int inst,
 
 u32 dp_mod_hash(struct module *owner, u16 ep)
 {
-	unsigned long index;
-
-	index = (unsigned long)owner;
-	return (u32)(((index >> DP_MOD_HASH_SHIFT) % DP_MOD_HASH_SIZE) | ep);
+	return hash_ptr(owner, DP_MOD_HASH_BIT_LENGTH);
 }
 
 struct dp_mod *dp_mod_lookup(struct hlist_head *head, struct module *owner,
diff --git a/drivers/net/datapath/dpm/datapath_instance.h b/drivers/net/datapath/dpm/datapath_instance.h
index 3ecc4c623c41..22d86ccbb6d6 100644
--- a/drivers/net/datapath/dpm/datapath_instance.h
+++ b/drivers/net/datapath/dpm/datapath_instance.h
@@ -13,13 +13,11 @@
 extern int dp_cap_num;
 extern struct dp_hw_cap hw_cap_list[DP_MAX_HW_CAP];
 
-#define DP_DEV_HASH_SHIFT 8
 #define DP_DEV_HASH_BIT_LENGTH 10
-#define DP_DEV_HASH_SIZE ((1 << DP_DEV_HASH_BIT_LENGTH) - 1)
+#define DP_DEV_HASH_SIZE (1 << DP_DEV_HASH_BIT_LENGTH)
 
-#define DP_MOD_HASH_SHIFT 8
 #define DP_MOD_HASH_BIT_LENGTH 10
-#define DP_MOD_HASH_SIZE ((1 << DP_MOD_HASH_BIT_LENGTH) - 1)
+#define DP_MOD_HASH_SIZE (1 << DP_MOD_HASH_BIT_LENGTH)
 
 #define NO_NEED_PMAC(flags)  ((flags & \
 		(DP_F_FAST_WLAN | DP_F_FAST_DSL)) && \
diff --git a/drivers/net/datapath/dpm/datapath_swdev.c b/drivers/net/datapath/dpm/datapath_swdev.c
index 594cda5125ae..db2ad0c16969 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.c
+++ b/drivers/net/datapath/dpm/datapath_swdev.c
@@ -41,11 +41,13 @@ static inline void swdev_unlock(void)
 
 u16 dp_swdev_cal_hash(unsigned char *name)
 {
-	size_t hash = 5381;
+	unsigned long hash = init_name_hash(0);
 
-	while (*name)
-		hash = 33 * hash ^ (unsigned char)*name++;
-	return (u16)(hash & 0x3F);
+	if (name) {
+		while (*name)
+			hash = partial_name_hash(*name++, hash);
+	}
+	return hash_long(hash, BRIDGE_ID_ENTRY_HASH_LENGTH);
 }
 
 /* dp_kfree - used as destructor for
diff --git a/drivers/net/datapath/dpm/datapath_swdev.h b/drivers/net/datapath/dpm/datapath_swdev.h
index 42c9d70a545c..2c9d0c17b66e 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.h
+++ b/drivers/net/datapath/dpm/datapath_swdev.h
@@ -10,12 +10,8 @@
 #define ENABLE 1
 #define DISABLE 0
 
-#define BRIDGE_ID_ENTRY_HASH_SHIFT			0
-#define BRIDGE_ID_ENTRY_HASH_LENGTH			6
-#define BRIDGE_ID_ENTRY_HASH_MASK ((BIT(BRIDGE_ID_ENTRY_HASH_LENGTH)) - 1)
+#define BRIDGE_ID_ENTRY_HASH_LENGTH 6
 #define BR_ID_ENTRY_HASH_TABLE_SIZE BIT(BRIDGE_ID_ENTRY_HASH_LENGTH)
-#define BRIDGE_ID_ENTRY_HASH_TABLE_VALUE(x) \
-	(((u32)(x) >> BRIDGE_ID_ENTRY_HASH_SHIFT) & BRIDGE_ID_ENTRY_HASH_MASK)
 
 #define ADD_BRENTRY	0x00000001
 #define DEL_BRENTRY	0x00000002
